### PR TITLE
feat(competition): public read endpoints for finalized rounds

### PR DESCRIPTION
**Pull request by whoabuddy**
## Summary

Adds four public, no-auth GET endpoints that let agents and tooling introspect finalized competition round data. Closes the gap left by the prior quest (2026-05-20-competition-snapshot-finalize) where the finalization system was admin-only.

- `GET /api/competition/rounds` — paginated list of finalized rounds (?limit 1-100 default 20, ?offset)
- `GET /api/competition/rounds/[roundId]` — full detail with all agent results ranked by P&L + reward rows; 404 on non-finalized
- `GET /api/competition/rounds/[roundId]/results/[stxAddress]` — per-agent permalink; 400 on invalid address, 404 on no placement
- `GET /api/competition/status?address=...` — extended with optional `latestRoundResult` when agent has a finalized placement

All four endpoints: no auth, self-document on `?docs=1`, return 404 for in-flight rounds (open/closed/finalizing only visible to admin).

## What's in this PR

**Phase 1** (commits a5b2ee9, 60900c3): D1 read helpers in `lib/competition/finalize/read.ts` + 28 unit tests

**Phase 2** (commits f7d7a01, 8a9f19f, 14c4c44): Four routes + status extension + 27 route tests. 319 competition tests pass.

**Phase 3** (this PR's final commits): Docs + OpenAPI

- `CLAUDE.md` — "Public Read Endpoints" subsection added to Competition Finalize section
- `app/llms-full.txt/route.ts` — Round Finalization section extended with four endpoint signatures and curl examples
- `app/docs/[topic]/route.ts` — competition-finalize topic doc: replaced stale "There are no new agent-facing endpoints" with full message formats for all four routes (matching bounties topic depth)
- `app/api/openapi.json/route.ts` — Three new path entries + `latestRoundResult` on status response

## Test plan

- [ ] `npm test` passes on this branch (319 competition tests, 28 read helper tests, 27 route tests all green)
- [ ] `npm run lint` passes (no errors; pre-existing `` warnings only)
- [ ] `curl https://aibtc.com/api/competition/rounds` — after deploy, should return live round list
- [ ] `curl "https://aibtc.com/api/competition/rounds/week-1-2026-05-13"` — should return detail for live test fixture
- [ ] `curl "https://aibtc.com/docs/competition-finalize.txt"` — should no longer say "no agent-facing endpoints"
- [ ] OpenAPI spec at `/api/openapi.json` includes the three new paths

## Quest context

- Quest dir: `.planning/2026-05-20-competition-rounds-read-endpoints/`
- Parent quest: `.planning/2026-05-20-competition-snapshot-finalize/` (archived)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by arc0btc — APPROVED** [review]
Adds four public read endpoints for finalized competition data — exactly what agents need to introspect round history without admin access. The phased approach (D1 helpers → routes → docs/OpenAPI) is clear and the diff tells a coherent story.

**What works well:**
- In-flight round leak guard is correct and well-tested — `getFinalizedRound` returns null for non-visible statuses, routes respond 404, and the tests explicitly verify no results leak (`expect(getRoundResults).not.toHaveBeenCalled()`).
- The `limit + 1` trick for `hasMore` detection avoids a COUNT query — clean.
- `mapRoundResult` correctly isolates the TEXT→RoundResult conversion so `JSON.parse` never touches `result_json` directly.
- Detail route correctly parallelizes `getRoundResults` + `getRoundRewards` with `Promise.all` after the round-existence gate.
- Status extension graceful degradation is solid: if `getLatestFinalizedRoundResultForAgent` throws, base status still returns 200 with full existing fields. Tests cover this path explicitly.
- Test suite is thorough: 400/404/429/503 paths, pagination edge cases (hasMore=true, limit clamping, negative offset), and the regression test verifying all prior status fields survive the extension.
- `CACHE_INVARIANTS:POSTURE=public-only-get` header comment is consistent with repo conventions.

**[suggestion] Parallelize status + latestRoundResult fetch** (`app/api/competition/status/route.ts`)

`getCompetitionStatusFromD1` and `getLatestFinalizedRoundResultForAgent` are independent D1 queries issued sequentially. Under D1 latency (~5–15ms each), this adds unnecessary serial wait on every status call. The graceful degradation pattern still works cleanly with `.catch()`:

```suggestion
    const [status, latestRoundResult] = await Promise.all([
      getCompetitionStatusFromD1(db, address),
      getLatestFinalizedRoundResultForAgent(db, address).catch((err) => {
        logger.warn("latestRoundResult lookup failed — omitting from response", {
          error: String(err),
          address,
        });
        return null;
      }),
    ]);
```

This keeps the outer catch for `getCompetitionStatusFromD1` failures (→ 503) while still degrading gracefully on `latestRoundResult` errors.

**[nit]** `VISIBLE_STATUSES` is interpolated as a string into SQL rather than bound — correct given D1 doesn't support binding array params for `IN` clauses, but the comment in `read.ts` doesn't say that. A one-liner noting "D1 does not support binding array params for IN" would clarify why the interpolation is intentional.

**[question]** The per-agent permalink (`rounds/[roundId]/results/[stxAddress]`) validates the STX address before hitting D1, which is right. The list route (`/rounds`) and detail route (`/rounds/[roundId]`) don't validate `roundId` format — is that intentional? D1 will return null on a garbage string, so it's safe, but a quick format check (non-empty, reasonable length) would make the 404 response cleaner for callers passing malformed IDs.

**Operational context:** We've been consuming competition round data through the admin surface in Arc's sensors. These public endpoints reduce coupling and are exactly what the wider agent ecosystem needs for post-round analysis. The `latestRoundResult` extension to `/status` is particularly useful — it gives agents their last placement without requiring a separate round-list call.

55 tests, all green. No critical performance violations, no blocking accessibility findings. Ship it.

File: lib/competition/finalize/read.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+ *   GET /api/competition/rounds
+ *   GET /api/competition/rounds/[roundId]
+ *   GET /api/competition/rounds/[roundId]/results/[stxAddress]
+ *   GET /api/competition/status?address=... (latestRoundResult extension)
+ *
+ * Routes stay thin — all SQL lives here.
+ *
+ * Visibility filter: all read helpers restrict to rounds with status in
+ * ('finalized', 'partially_paid', 'paid'). In-flight rounds (open, closed,
+ * finalizing) are excluded from public read surfaces.
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply lib/competition/finalize/read.ts]: Fixed in bcafc4a — clarified the file header to specify that only listFinalizedRounds, getFinalizedRound, and getLatestFinalizedRoundResultForAgent enforce the visibility filter in SQL. The result/reward helpers (getRoundResults / getRoundRewards / getRoundResultForAgent) intentionally do NOT join against competition_rounds — they're guarded by routes calling getFinalizedRound first as a gate. Header now points to the gate pattern used by the route files.

File: app/api/competition/rounds/route.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    limit = Math.min(Math.max(parsed, MIN_LIMIT), MAX_LIMIT);
+  }
+
+  // Parse and validate offset
+  const offsetParam = searchParams.get("offset");
+  let offset = 0;
+  if (offsetParam !== null) {
+    const parsed = parseInt(offsetParam, 10);
+    offset = Math.max(Number.isFinite(parsed) ? parsed : 0, 0);
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/competition/rounds/route.ts]: Fixed in bcafc4a — invalid or negative offset now returns 400 (matching the limit behavior and the OpenAPI spec). Updated route + tests (replaced the offset=-1 coerce-to-0 test with a 400 expectation, added a non-integer offset=abc test).

File: app/api/openapi.json/route.ts [hunk 3]
```diff
@@ -1081,6 +1081,48 @@ export function GET() {
...
                       first_trade_at: { type: ["integer", "null"], description: "Unix seconds" },
                       last_trade_at: { type: ["integer", "null"], description: "Unix seconds" },
+                      latestRoundResult: {
+                        description:
+                          "The agent's result in the most recent finalized round. " +
+                          "Only present when the agent has a placement in at least one " +
+                          "finalized round (status: finalized, partially_paid, or paid). " +
+                          "Omitted when the agent has no round placements.",
+                        type: "object",
+                        nullable: true,
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/openapi.json/route.ts]: Fixed in bcafc4a — removed `nullable: true` from the latestRoundResult schema and tightened the description to say the field is omitted from the response object entirely when absent (never present with a null value). Matches the actual wire behavior.

File: app/docs/[topic]/route.ts [hunk 4]
```diff
@@ -985,19 +985,174 @@ Results are immutable once written. The admin route is at
...
+      "unpriced_tokens": []
+    },
+    "calculated_at": "2026-05-20T12:34:56.789Z"
+  }
+}
+\`\`\`
+
+Response \`400\` — invalid STX address:
+\`\`\`json
+{ "error": "Invalid stxAddress path parameter. Expected a Stacks mainnet address (SP… / SM…)." }
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/docs/[topic]/route.ts]: Fixed in bcafc4a — the docs 400 example for stxAddress now includes the `example` field that the route actually returns, so callers see the full shape.

File: app/api/competition/status/route.ts [hunk 5]
```diff
@@ -112,7 +117,27 @@ export async function GET(request: NextRequest) {
...
+    // Optional extension: add the agent's latest finalized round result when available.
+    // Graceful degradation: if this call fails, the base status response is still returned.
+    let latestRoundResult = null;
+    try {
+      latestRoundResult = await getLatestFinalizedRoundResultForAgent(db, address);
+    } catch (err) {
+      logger.warn("latestRoundResult lookup failed — omitting from response", {
+        error: String(err),
+        address,
+      });
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/competition/status/route.ts]: Acknowledged — partially mitigated in bcafc4a by parallelizing the two D1 reads via Promise.all (per arc0btc's review). Wall-time is now bounded by the slower query rather than their sum, which removes most of the latency cost for both registered and unregistered callers. Skipping the lookup entirely on `registered: false` would save one cheap indexed read per probe, but it would also re-serialize the calls (we don't know `registered` until status returns). Leaving the conditional skip out for now; happy to add it as a follow-up if /status traffic from unregistered probes becomes meaningful in worker-logs.

File: app/api/openapi.json/route.ts [hunk 6]
```diff
@@ -2182,6 +2224,369 @@ export function GET() {
...
+              },
+            },
+            "400": {
+              description: "Invalid limit or offset parameter",
+              content: {
+                "application/json": {
+                  schema: { $ref: "#/components/schemas/ErrorResponse" },
+                },
+              },
+            },
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/openapi.json/route.ts:2310]: Fixed in bcafc4a — the implementation side moved to match the spec. Invalid offset now returns 400 (same as limit), so the OpenAPI 400 description "Invalid limit or offset parameter" is now accurate without further edits.

File: lib/competition/finalize/read.ts [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    SELECT
+      crr.round_id, crr.rank, crr.stx_address, crr.btc_address, crr.erc8004_agent_id,
+      crr.trade_count, crr.priced_trade_count, crr.unpriced_trade_count,
+      crr.volume_usd, crr.received_usd, crr.pnl_usd, crr.pnl_percent,
+      crr.latest_trade_at, crr.result_json, crr.calculated_at
+    FROM competition_round_results crr
+    JOIN competition_rounds cr ON cr.round_id = crr.round_id
+    WHERE crr.stx_address = ?1
+      AND cr.status IN ${VISIBLE_STATUSES}
+    ORDER BY cr.starts_at DESC
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply lib/competition/finalize/read.ts:292]: Acknowledged but deferred — adding the `(stx_address, round_id)` index is the right call, but this quest (2026-05-20-competition-rounds-read-endpoints) has "no new migrations" as an explicit scope-out (the table came from migration 017 in the prior quest). The current table is small (12 rows for week-1) so the scan cost is negligible right now; will file a follow-up issue for the index as part of the next competition migration batch. Parallelizing /status (bcafc4a) reduces the immediate hot-path concern in the meantime.
